win_firewall_rule: Implement idempotency, check-mode and diff support

dagwieers opened the pull request:
##### SUMMARY
This PR includes the following changes:
- An idempotency fix when `profile: any`
- Documentation fixes (remove `required: false`)
- Parameter handling fixes
- RDP example that matches default RDP rule
- Renamed parameter 'enable' to 'enabled' (kept alias for compatibility)
- Renamed parameter 'profile' to 'profiles' (kept alias for compatibility)
- Implemented check-mode support
- Implemented diff support
- Adapted integration tests 
- Returns proper command output on failure.

This fixes #23455 and #24976

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_firewall_rule

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

dagwieers committed:
win_firewall_rule: Small idempotency fix

This PR includes the following changes:
- an idempotency fix when `profile: any`
- better difference output to debug idempotency issues
- documentation fixes (remove `required: false`)
- Parameter handling fixes
- RDP example that matches default RDP rule
- Renamed parameter 'enable' to 'enabled' (kept alias)
- Renamed parameter 'profile' to 'profiles' (kept alias)

mattclay commented:
Integration test failure: https://app.shippable.com/github/ansible/ansible/runs/17791/9/tests

Here's one example from `test/integration/targets/win_firewall_rule/tasks/main.yml:39`:

```
{
    "assertion": "add_firewall_rule_again.changed == false", 
    "changed": false, 
    "evaluated_to": false, 
    "failed": true
}
```

ar7z1 commented on lib/ansible/modules/windows/win_firewall_rule.ps1:
```diff
@@ -291,7 +282,12 @@ $fwsettings.Add("LocalPort", $localport);
+If ($profiles -eq "any") {
+    $fwsettings.Add("Profiles", "Domain,Private,Public")
+} Else {
+    $fwsettings.Add("Profiles", (Get-Culture).textinfo.totitlecase($profiles))
```
It seems it can be removed. Comparison operator `-eq` are [case-insensitive by default](https://msdn.microsoft.com/en-us/powershell/reference/5.1/microsoft.powershell.core/about/about_comparison_operators):
>By default, all comparison operators are case-insensitive. To make a comparison operator case-sensitive, precede the operator name with a `c`. For example, the case-sensitive version of `-eq` is `-ceq`. To make the case-insensitivity explicit, precede the operator with an `i`. For example, the explicitly case-insensitive version of `-eq` is `-ieq`.

dagwieers replied:
Aha, good to know. Thanks !

ar7z1 reviewed (approved):
shipit

ar7z1 commented:
@dagwieers Thank you for excellent work!
I can't understand what's wrong with the tests. I will have access to my dev machine in a few days. I'll try to debug the failed tests.

jborean93 commented:
shipit

dagwieers commented:
@ar7z1 I still need to look at the test failures. At a first look I couldn't find why my changes would have broken them.

There are also some other things I would like to do:
- Remove trailing semi-colons
- Some more cosmetic consistency changes
- Use normal dicts instead of PSObject (for $result)
- Implement check-mode support

But too many changes in a single PR is frowned upon, and cosmetic changes double-frowned.

ar7z1 commented:
@dagwieers I created a [PR](https://github.com/dagwieers/ansible/pull/5) to your PR. ;-)

dagwieers commented:
@ar7z1 I have retained the fix for the integration tests from your PR. So that this PR becomes mergeable again.

However I noticed my changes to "differences" do not make sense (useful for debugging, but output is not very clear), and do not use the normal provisions for showing diff output (-D/--diff). So we definitely need to revise this too. So I am going to fix my PR for that too.

dagwieers committed:
Rewrite module completely

The logic is still intact, but various changes with a single goal:

- Make the module idempotent
- Implement check-mode
- Implement diff-mode
- Adapted integration tests

This fixes #18807 and #23455.

dagwieers commented:
Alright, after changes what needed to be changed I ended up rewriting most of the module.
Although the logic is still the same.

lfrancke commented on lib/ansible/modules/windows/win_firewall_rule.ps1:
```diff
@@ -59,346 +77,380 @@ function preprocessAndCompare($key, $outputValue, $fwsettingValue) {
+            $result.msg += @("Created firewall rule '$name'")
+        } else {
+            $result.failed = $true
+            $result.msg += @("Create command '$command' failed with rc=$rc")
```
I'm a) new to PowerShell b) to the Ansible internals and c) Windows so disregard if needed but will this retain the error message somehow that netsh returned? That's helpful for debugging.

dagwieers replied:
Yes, $result.output is retained on a non-zero return code (later in the code).

dagwieers replied:
@lfrancke I think ansibot is waiting for me to fix this, but there is nothing to fix. Can you approve your review ?

lfrancke replied:
Uhhmm I'm sorry for the hold-up. I have to admit I have no idea how to approve a review here?

dagwieers replied:
I don't know *if* you are holding things up, but ansibot says: `waiting_on: maintainer` and I don't know why honestly. So one tends to speculate ;-)

So what you do is go to the *Files changed* tab of the PR, then at the top click *Review changes* and select *Approve*.

lfrancke replied:
Thanks! Done.

lfrancke commented:
As commented on #23455 I did test this PR for my specific use-case and it fixes all the issues I had with the win_firewall_rule module. Thank you @dagwieers !

dagwieers commented:
I was lucky I was refactoring the module when your report came in ;-) Thanks for testing !

dagwieers committed:
Change casing to lowercase

dagwieers commented:
@nitzmahone All fixed, ready for your re-review.

jborean93 commented:
I can't say I know exactly what is changing here but if the tests still pass and what I have briefly seen we should get this through so people can start seeing the effects in devel.

shipit

ar7z1 commented on lib/ansible/modules/windows/win_firewall_rule.ps1:
```diff
@@ -20,12 +20,30 @@
+    $output = ""
+    foreach($item in $object.GetEnumerator() | Sort -Property Name) {
+        if (($item.Name -notin $unsupported) -and ($item.Value -ne $null)) {
+            $output += "$($item.Name): $($item.Value)`n"
```
Did you mean `\n`?

dagwieers replied:
Well, it's powershell, and `n is how it's done in PS.

ar7z1 replied:
Oh, you are right! :-)

ar7z1 commented on lib/ansible/modules/windows/win_firewall_rule.ps1:
```diff
@@ -54,351 +72,390 @@ function preprocessAndCompare($key, $outputValue, $fwsettingValue) {
+                        $result.diff.after = ConvertTo-SortedKV $fwsettings
+                        $result.diff.before = ConvertTo-SortedKV $rule $unsupported
+                    }
+                    ForEach($rule in $output.GetEnumerator()) {
```
What do you think about move `if` outside and add `break` statement?

```powershell
:outer
ForEach($rule in $output.GetEnumerator()) {
    ForEach($fwsetting in $fwsettings.GetEnumerator()) {
        if ($rule.$fwsetting -ne $fwsettings.$fwsetting) {
            $diff = $true
            break outer
        }
    }
}

if (-not $diff) {
    $result.identical = $true
}
```

dagwieers replied:
Well, I mostly translated the existing code to the newer structure, but now looking at what we are doing, it is no longer needed to do this. We can simply compare before and after to see if there is a difference.

I will fix this.

ar7z1 commented on lib/ansible/modules/windows/win_firewall_rule.ps1:
```diff
@@ -54,351 +72,390 @@ function preprocessAndCompare($key, $outputValue, $fwsettingValue) {
+                    $result.output = $HashProps
+                }
+        }
+        if ($rc -eq 0 -or $rc -eq 1) {
```
Does exit code `1` means "rule not found"?

dagwieers replied:
To be honest, I don't know or remember :-/
However, the integration tests do remove the same rule twice to cover this situation, and it is doing the right thing.

ar7z1 commented on lib/ansible/modules/windows/win_firewall_rule.ps1:
```diff
@@ -54,351 +72,390 @@ function preprocessAndCompare($key, $outputValue, $fwsettingValue) {
+                $result.diff.after = ""
+                $result.diff.before = ConvertTo-SortedKV $fwsettings
+            }
+            $result.changed = $true
```
I think we should set `chaged` to `$false` when rule does not exist.

ar7z1 commented on test/integration/targets/win_firewall_rule/tasks/main.yml:
```diff
@@ -3,98 +3,94 @@
 - name: Check that creating new firewall rule succeeds with a change
   assert:
     that:
-    - add_firewall_rule.failed == false
```
I think it will be more safety if we leave such checks. :)

dagwieers replied:
The check does not buy us anything, if it would have failed we would not be doing the assert. (Unless we used `ignore_errors: yes`).

ar7z1 commented:
Definitely shipit!

dagwieers committed:
Improve the logic wrt. diff

ar7z1 commented:
shipit

dagwieers commented:
ansibot

jborean93 commented:
Hopefully this will also give it a nudge

shipit

dagwieers commented:
bot_status

dagwieers commented:
We need more maintainers, or someone from the core team to get this merged.

skasai commented:
Question - Would this be the reason I can't set/update firewall rules with current win_firewall_rule?  I am currently dipping my feet into Ansible and noticing when I did tests on machines, I wasn't creating the rule.  Using the **force** option did delete pre-defined rule.

dagwieers commented:
@skasai Yup, as I added to the notes in this PR:

> - Modifying existing firewall rules is not possible, the module does allow replacing complete rules based on name, but that works by removing the existing rule completely, and recreating it with provided information (when using C(force)).

dagwieers commented:
@nitzmahone Can you please merge this PR ? It is almost 2 months old, and people are still having issues with things that this PR fixes.

skasai commented:
> Yup, as I added to the notes in this PR:

Modifying existing firewall rules is not possible, the module does allow replacing complete rules based on name, but that works by removing the existing rule completely, and recreating it with provided information (when using C(force)).

Ok...  I was playing around with this last night and today and scratching my head as to why it said changed and looking on the test machines, I see no added rule and wondering why it deleted the existing rule when I did force. 

Even running with -vvvvvv, wasn't really indicative of what was going on.

dagwieers commented:
@skasai Was this with PR #23162 or with the version in v2.3 or devel ?

skasai commented:
At the moment, not sure, since I know the version of Ansible is v2.3.0.0 right now and what I see near the end is this:

changed: [<machine>] => {
    "changed": true, 
    "difference": [], 
    "failed": false, 
    "fwsettings": {
        "Action": "allow", 
        "Direction": "in", 
        "Enabled": "yes", 
        "LocalIP": "any", 
        "LocalPort": "any", 
        "Profiles": "Private,Public", 
        "Protocol": "ICMPv4", 
        "RemoteIP": "<iprange>/24", 
        "RemotePort": "any", 
        "Rule Name": "ATest", 
        "Service": "any"
    }, 
    "msg": [
        "No rule could be found", 
        "Created firewall rule ATest"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************
<machine>       : ok=2    changed=1    unreachable=0    failed=0   

This is on CentOS 7 doing a yum install so not devel.  Not sure how to do PR or what not.

dagwieers commented:
@skasai If you would have been running this code, it would have told you that you cannot provide a `localport` or `remoteport` when using a protocol other than TCP or UDP.

skasai commented:
Figured as much.  How would one pull down a PR then for testing?  I saw exerpt code for files, but didn't seem to line up so I avoided trying to hand modify.

dagwieers commented:
@skasai The easiest option is to take the file and put it in *library/*, however in this case we also modify the vmware library, so that's not going to work. If you don't know git very well, maybe it's best to wait for this to be released. At least you now know how to fix this :-)

skasai commented:
Took me a while and a few mistakes but hand modified it based off the PR mods and it looks like this works.  Was able to create the test firewall rule when earlier behavior showed no changes.

Do note that it treated 'existing rule' with 'force' as not counting as 'changed', but I suppose this is expected behavior?

dagwieers commented:
Indeed, it compares the before and after, and only reports a change if there is a difference. Thanks for testing !

skasai commented:
@dagwieers Ok.  Although technically if you are forcing a change in values on an existing rule, technically, that is a change. :)

What I see is this:

    "msg": [
        "The rule 'File and Printer Sharing (Echo Request - ICMPv4-In)' exists.", 
        "The rule 'File and Printer Sharing (Echo Request - ICMPv4-In)' exists but has different values", 
        "Removed the rule 'File and Printer Sharing (Echo Request - ICMPv4-In)'", 
        "Created firewall rule 'File and Printer Sharing (Echo Request - ICMPv4-In)'"
    ]

Run again:
    "msg": [
        "The rule 'File and Printer Sharing (Echo Request - ICMPv4-In)' exists.", 
        "An identical rule exists", 
        "Firewall rule 'File and Printer Sharing (Echo Request - ICMPv4-In)' was already created"
    ]

In both cases, Ansible did show 'changed': false.
ok: ['machine'] => {
    "changed": false, 
    "diff": null, 
    "fwsettings": {
        "Action": "Allow", 
        "Direction": "In", 
        "Edge traversal": "No", 
        "Enabled": "Yes", 
        "InterfaceTypes": "Any", 
        "LocalIP": "any", 
        "Profiles": "Private,Public", 
        "Protocol": "ICMPV4", 
        "RemoteIP": "'iprange'/24", 
        "Rule Name": "File and Printer Sharing (Echo Request - ICMPv4-In)", 
        "Security": "NotRequired"
    },

dagwieers commented:
@skasai That's a bug.
